src/rbm.jl

```diff
@@ -1,118 +1,107 @@
-function BernoulliRBM(n_vis::Int, n_hid::Int; sigma=0.001, momentum=0.9,
-        decay_rate=0.0, sparsity_cost=0.0, sparsity_exp=0.5)
-    RBM(
-        Bernoulli, Bernoulli, n_vis, n_hid;
-        sigma=sigma, momentum=momentum, decay_rate=decay_rate,
-        sparsity_cost=sparsity_cost, sparsity_exp=sparsity_exp
-    )
+@runonce type RBM{T,V,H} <: AbstractRBM{T,V,H}
+    W::Matrix{T}         # matrix of weights between vis and hid vars
+    vbias::Vector{T}     # biases for visible variables
+    hbias::Vector{T}     # biases for hidden variables
 end
 
-typealias GRBM RBM{Gaussian, Bernoulli}
-function GRBM(n_vis::Int, n_hid::Int; sigma=0.001, momentum=0.9,
-        decay_rate=0.0, sparsity_cost=0.0, sparsity_exp=0.5)
-    RBM(
-        Gaussian, Bernoulli, n_vis, n_hid;
-        sigma=sigma, momentum=momentum, decay_rate=decay_rate,
-        sparsity_cost=sparsity_cost, sparsity_exp=sparsity_exp
-    )
+function RBM(T::Type, V::Type, H::Type,
+             n_vis::Int, n_hid::Int; sigma=0.01)
+    RBM{T,V,H}(map(T, rand(Normal(0, sigma), (n_hid, n_vis))),
+             zeros(n_vis), zeros(n_hid))
 end
 
+RBM(V::Type, H::Type, n_vis::Int, n_hid::Int; sigma=0.01) =
+    RBM(Float64, V, H, n_vis, n_hid; sigma=sigma)
+
 
-function logistic(x)
-    return 1 ./ (1 + exp(-x))
+# some well-known RBM kinds
+BernoulliRBM(n_vis::Int, n_hid::Int; sigma=0.01) =
+    RBM(Float64, Degenerate, Bernoulli, n_vis, n_hid; sigma=sigma)
+
+GRBM(n_vis::Int, n_hid::Int; sigma=0.01) =
+    RBM(Float64, Normal, Bernoulli, n_vis, n_hid; sigma=sigma)
+
+
+function Base.show{T,V,H}(io::IO, rbm::RBM{T,V,H})
+    n_vis = size(rbm.vbias, 1)
+    n_hid = size(rbm.hbias, 1)
+    print(io, "RBM{$V,$H}($n_vis, $n_hid)")
 end
 
+## utils
 
-function hid_means(rbm::RBM, vis::Mat{Float64})
+function hid_means{T}(rbm::RBM, vis::Mat{T})
     p = rbm.W * vis .+ rbm.hbias
     return logistic(p)
 end
 
 
-function vis_means(rbm::RBM, hid::Mat{Float64})
+function vis_means{T}(rbm::RBM, hid::Mat{T})
     p = rbm.W' * hid .+ rbm.vbias
     return logistic(p)
 end
 
 
-function sample(::Type{Bernoulli}, means::Mat{Float64})
-    return float(rand(size(means)) .< means)
+## samping
+
+function sample{T}(::Type{Degenerate}, means::Mat{T})
+    return means
 end
 
+function sample{T}(::Type{Bernoulli}, means::Mat{T})
+    return map(T, float((rand(size(means)) .< means)))
+end
 
-function sample(::Type{Gaussian}, means::Mat{Float64})
+
+function sample{T}(::Type{Gaussian}, means::Mat{T})
     sigma2 = 1                   # using fixed standard diviation
-    samples = zeros(size(means))
+    samples = zeros(T,size(means))
     for j=1:size(means, 2), i=1:size(means, 1)
-        samples[i, j] = rand(Normal(means[i, j], sigma2))
+        samples[i, j] = T(rand(Normal(means[i, j], sigma2)))
     end
     return samples
 end
 
 
-function sample_hiddens{V,H}(rbm::AbstractRBM{V, H}, vis::Mat{Float64})
+function sample_hiddens{T,V,H}(rbm::AbstractRBM{T,V,H}, vis::Mat{T})
```

Comment:
If we agree on implicit conversion, I don't see a reason to limit `T` to be an instance of `AbstractFloat` - one may be interested in passing matrix of `Int`s or even `Boolean`, which is very natural for Bernoulli RBM.